src/test/java/pl/karolpat/service/ParkingMeterServiceTest.java

```diff
@@ -0,0 +1,194 @@
+		start = new Timestamp(calendar.getTimeInMillis());
+
+	}
+
+	@Test
+	public void testSave() {
+		// given
+		when(parkingMeterRepo.save(parkingMeter1)).thenReturn(parkingMeter1);
+		// when
+		ParkingMeter result = parkingMeterService.save(parkingMeter1);
+		// then
+		assertThat(result).isInstanceOf(ParkingMeter.class);
+		assertEquals(parkingMeter1, result);
+	}
+
+	@Test
+	public void testFindUserParkingMeter() {
+
+		user1.setId(1l);
+		parkingMeter1.setStart(Timestamp.valueOf("2017-1-1 12:55:00"));
+		parkingMeter2.setStart(Timestamp.valueOf("2017-1-1 12:55:30"));
+		parkingMeter1.setUser(user1);
+		parkingMeter2.setUser(user1);
+
+		when(parkingMeterRepo.getFirstByUserIdOrderByStart(1l)).thenReturn(parkingMeter2);
+
+		ParkingMeter result = parkingMeterService.findUserParkingMeter(1l);
+
+		assertThat(result).isInstanceOf(ParkingMeter.class);
+		assertEquals(parkingMeter2, result);
+	}
+
+	@Test
+	public void testFindAll() {
+
+		List<ParkingMeter> list = Arrays.asList(parkingMeter1, parkingMeter2);
+		when(parkingMeterRepo.findAll()).thenReturn(list);
+
+		List<ParkingMeter> result = parkingMeterService.findAll();
+
+		assertThat(result).hasOnlyElementsOfType(ParkingMeter.class);
+		assertEquals(list, result);
+	}
+
+	@Test
+	public void testCheckCost_when_users_is_vip() {
+
+		parkingMeter1.setStart(start);
+
+		user1.setId(1);
+		user1.setVip(true); // User's vip status to true
+
+		Set<ParkingMeter> parkings = new HashSet<>();
+		parkings.add(parkingMeter1);
+		user1.setParking(parkings);
+
+		when(parkingMeterRepo.getFirstByUserIdOrderByStart(1l)).thenReturn(parkingMeter1);
+
+		Map<String, Double> map = parkingMeterService.checkCost(user1);
+
+		assertNotNull(map);
+		assertThat(map).containsKey("PLN");
+		assertThat(map).containsValue(16.25);
+		assertThat(map).containsKey("Hours spent");
+		assertThat(map).containsValue(5.0);
+	}
+
+	@Test
+	public void testCheckCost_when_users_is_NOT_vip() {
+
+		parkingMeter1.setStart(start);
+
+		user1.setId(1);
+		user1.setVip(false); // User's vip status to false
+
+		Set<ParkingMeter> parkings = new HashSet<>();
+		parkings.add(parkingMeter1);
+		user1.setParking(parkings);
+
+		when(parkingMeterRepo.getFirstByUserIdOrderByStart(1l)).thenReturn(parkingMeter1);
+
+		Map<String, Double> map = parkingMeterService.checkCost(user1);
+
+		assertNotNull(map);
+		assertThat(map).containsKey("PLN");
+		assertThat(map).containsValue(31.0);
+		assertThat(map).containsKey("Hours spent");
+		assertThat(map).containsValue(5.0);
+	}
+
+	@Test
+	public void testGetCurrentHours() {
+
+		parkingMeter1.setStart(start);
+		when(parkingMeterRepo.getFirstByUserIdOrderByStart(1l)).thenReturn(parkingMeter1);
+
+		int result = parkingMeterService.getCurrentHours(1l);
+
+		assertEquals(5, result);
+	}
+
+	@Test
+	public void testGetCostIfVip() {
+
+		int hours = 5;
+
+		double result = parkingMeterService.getCostIfVip(hours);
+		assertEquals(16.25, result, 0);
```

Comment:
epsilon na 0 - czy tak powinno się robić?

Reply:
Wydawało mi się, że epsilon odpowiada za dopuszczalne odchylenie między porównywanymi wartościami i przyjąłem, że jeśli dwie liczby mają być dokładnie równe sobie, to ta różnica powinna być jak najmniejsza. Czy w takim razie powinienem użyć niezerowej wartości równej oczekiwanej dokładności, np. 0,001?